### PR TITLE
[ALLUXIO-3069] Async partial caching

### PR DESCRIPTION
**Pull request by calvinjia**
https://alluxio.atlassian.net/browse/ALLUXIO-3069

Points of interest

* Client side: `FileInStream` changed to only read data and send async cache requests to workers. Caching of blocks is now fully handled by the worker.
* Client side: `InStreamOptions` is not longer a traditional `options` class since it is not user facing
* Protocol: Added a new request, `AsyncCacheRequest` which is sent through `NettyRPC` from client to worker
* Worker side: Added `AsyncCacheManager` to handle scheduling, deduplicating, and executing async cache requests
* Worker side: Added a `RemoteBlockReader` to read blocks from other Alluxio workers

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java [hunk 1]
```diff
@@ -71,51 +74,72 @@
...
+      LOG.debug("Creating short circuit input stream for block {} @ {}", blockId, dataSource);
       try {
-        LOG.debug("Creating short circuit input stream for block {} @ {}", blockId, address);
-        return createLocalBlockInStream(context, address, blockId, blockSize, options);
+        return createLocalBlockInStream(context, dataSource, blockId, blockSize, options);
       } catch (NotFoundException e) {
         // Failed to do short circuit read because the block is not available in Alluxio.
-        // We will try to read from UFS via netty. So this exception is ignored.
-        LOG.warn("Failed to create short circuit input stream for block {} @ {}", blockId, address);
+        // We will try to read from via netty. So this exception is ignored.
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java]: remove "from"?

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java [hunk 2]
```diff
@@ -71,51 +74,72 @@
...
+
+    boolean shortCircuit = Configuration.getBoolean(PropertyKey.USER_SHORT_CIRCUIT_ENABLED);
+    boolean sourceSupportsDomainSocket = !NettyUtils.isDomainSocketSupported(dataSource);
+    boolean sourceIsLocal = dataSourceType == BlockInStreamSource.LOCAL;
+
+    // Short circuit
+    if (sourceIsLocal && shortCircuit && !sourceSupportsDomainSocket) {
+      LOG.debug("Creating short circuit input stream for block {} @ {}", blockId, dataSource);
       try {
-        LOG.debug("Creating short circuit input stream for block {} @ {}", blockId, address);
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java]: should we still keep this debug msg?

File: core/client/fs/src/main/java/alluxio/client/file/FileInStream.java [hunk 3]
```diff
@@ -521,402 +190,100 @@ private int positionedReadInternal(long pos, byte[] b, int off, int len) throws
...
+          try {
+            NettyRPCContext rpcContext =
+                NettyRPCContext.defaults().setChannel(channel).setTimeout(channelTimeout);
+            NettyRPC.call(rpcContext, new ProtoMessage(request));
+          } finally {
+            mContext.releaseNettyChannel(worker, channel);
+          }
+        }
+      } catch (Exception e) {
+        LOG.warn("Failed to complete async cache request: {}", e.getMessage());
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/FileInStream.java]: can you add block id and worker address to the warn message?

File: core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    } catch (AlluxioException | IOException e) {
+      // This is only best effort
+      LOG.warn("Failed to async cache block {} from UFS on copying the block: {}", blockId,
+          e.getMessage());
+      return false;
+    } finally {
+      try {
+        mBlockWorker.closeUfsBlock(Sessions.ASYNC_CACHE_SESSION_ID, blockId);
+      } catch (AlluxioException | IOException ee) {
+        LOG.warn("Failed to close UFS block {}", blockId);
```
Comment (apc999) [inline core/server/worker/src/main/java/alluxio/worker/block/AsyncCacheRequestManager.java]: add `ee.getMessage()` into warn msg?

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java (hunk 2 above)
Comment (calvinjia) [reply core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java]: The message is one line above outside the try block

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java (hunk 1 above)
Comment (calvinjia) [reply core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java]: Fixed, thanks.

File: core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java (hunk 2 above)
Comment (apc999) [reply core/client/fs/src/main/java/alluxio/client/block/stream/BlockInStream.java]: ack

**Comment by calvinjia** [comment]
@apc999 LGTM

**Pull request merged**